include/pybind11/cast.h

```diff
@@ -125,6 +243,158 @@ PYBIND11_NOINLINE inline handle get_type_handle(const std::type_info &tp, bool t
     return handle(type_info ? ((PyObject *) type_info->type) : nullptr);
 }
 
+struct value_and_holder {
+    instance *inst;
+    size_t vpos, index;
+    detail::type_info *type;
+
+    static constexpr size_t bits_per_ptr = 8 * sizeof(void *);
+
+    value_and_holder(instance *i, detail::type_info *type, size_t vpos, size_t index)
+        : inst{i}, vpos{vpos}, index{index}, type{type} {}
+
+    template <typename V = void> V *&value_ptr() const {
+        return reinterpret_cast<V *&>(
+                inst->simple_layout ? inst->simple_value_holder[0] :
+                inst->values_and_holders[vpos]);
+    }
+    // True if this `value_and_holder` has a non-null value pointer
+    operator bool() const { return value_ptr(); }
+
+    template <typename H> H &holder() const {
+        return reinterpret_cast<H &>(
+                inst->simple_layout ? inst->simple_value_holder[1] :
+                inst->values_and_holders[vpos + 1]);
+    }
+    bool holder_constructed() const {
+        if (inst->simple_layout) return inst->simple_holder_constructed;
+        return reinterpret_cast<std::uintptr_t>(inst->values_and_holders[index / bits_per_ptr])
```

Comment:
My thinking was to potentially save space, but given that it's likely to be any different until you inherit from more than 8 different pybind classes, that was perhaps premature optimization.  I'll simplify it to byte-level.